renterhost/contracts.go

```diff
@@ -0,0 +1,274 @@
+type Contract struct {
+	ID              types.ElementID
+	Revision        types.FileContract
+	HostSignature   types.Signature
+	RenterSignature types.Signature
+}
+
+// ValidateSignatures checks that the renter and host signatures are valid.
+func (c *Contract) ValidateSignatures(vc consensus.ValidationContext) (err error) {
+	hash := vc.ContractSigHash(c.Revision)
+	if c.Revision.HostPublicKey.VerifyHash(hash, c.HostSignature) {
+		err = fmt.Errorf("failed to validate host signature: %w", ErrInvalidSignature)
+	} else if c.Revision.RenterPublicKey.VerifyHash(hash, c.RenterSignature) {
+		err = fmt.Errorf("failed to validate renter signature: %w", ErrInvalidSignature)
+	}
+	return
+}
+
+// EncodeTo implements types.EncoderTo.
+func (c *Contract) EncodeTo(enc *types.Encoder) {
+	c.ID.EncodeTo(enc)
+	c.Revision.EncodeTo(enc)
+	c.HostSignature.EncodeTo(enc)
+	c.RenterSignature.EncodeTo(enc)
+}
+
+// DecodeFrom implements types.DecoderFrom.
+func (c *Contract) DecodeFrom(dec *types.Decoder) {
+	c.ID.DecodeFrom(dec)
+	c.Revision.DecodeFrom(dec)
+	c.HostSignature.DecodeFrom(dec)
+	c.RenterSignature.DecodeFrom(dec)
+}
+
+// MaxLen implements rpc.Object.
+func (c *Contract) MaxLen() uint64 {
+	return 10e3
+}
+
+// PaymentRevision returns a new file contract revision with the amount added to
+// the host payout fields and subtracted from the renter payout fields.
+func (c *Contract) PaymentRevision(amount types.Currency) (types.FileContract, error) {
+	if c.Revision.ValidRenterOutput.Value.Cmp(amount) < 0 || c.Revision.MissedRenterOutput.Value.Cmp(amount) < 0 {
+		return types.FileContract{}, errors.New("not enough funds")
+	}
+	fcr := c.Revision
+	fcr.RevisionNumber++
+	fcr.ValidHostOutput.Value = fcr.ValidHostOutput.Value.Add(amount)
+	fcr.MissedHostOutput.Value = fcr.MissedHostOutput.Value.Add(amount)
+	fcr.ValidRenterOutput.Value = fcr.ValidRenterOutput.Value.Sub(amount)
+	fcr.MissedRenterOutput.Value = fcr.MissedRenterOutput.Value.Sub(amount)
+	return fcr, nil
+}
+
+// ClearingRevision returns a new file contract revision with the revision
+// number set to uint64 max, the file fields set to their zero value, and the
+// missed proof outputs set to the valid proof outputs.
+func (c *Contract) ClearingRevision() types.FileContract {
+	fcr := c.Revision
+	fcr.RevisionNumber = math.MaxUint64
+	fcr.Filesize = 0
+	fcr.FileMerkleRoot = types.Hash256{}
+	fcr.MissedHostOutput.Value = fcr.ValidHostOutput.Value
+	fcr.MissedRenterOutput.Value = fcr.ValidRenterOutput.Value
+	return fcr
+}
+
+// ProgramRevision returns a new file contract revision with the burn amount
+// subtracted from the missed host output.
+func (c *Contract) ProgramRevision(burn types.Currency) (types.FileContract, error) {
+	if c.Revision.MissedHostOutput.Value.Cmp(burn) < 0 {
+		return types.FileContract{}, errors.New("not enough funds")
+	}
+	fcr := c.Revision
+	fcr.RevisionNumber++
+	fcr.MissedHostOutput.Value = fcr.MissedHostOutput.Value.Sub(burn)
```

Comment:
Instead of having a burn output, the missed proof outputs can't be greater than the sum of the valid proof outputs and the sum of the valid proof outputs must be the same as the parent contract.

Reply:
oh yeah. Forgot I did that lol